File: 01-reference-agent/java/src/main/java/AgentCore.java
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Domain logic for 01-reference-agent (no HTTP here).
 * Emits event maps: meta | token | error | metrics | done.
 */
public final class AgentCore {
    public static final List<Persona> PERSONAS = List.of(
            new Persona("conservative-charlie", "Conservative Charlie", "conservative"),
            new Persona("neutral-nancy", "Neutral Nancy", "neutral"),
            new Persona("thoughtless-toby", "Thoughtless Toby", "risk-taker")
    );

    private static final String CANNED_INPUT =
            "No ticker stories loaded yet. Ask the user to click Get Stories, "
                    + "then produce a brief placeholder note that you are waiting for headlines.";

    private static final String DEFAULT_BEDROCK_MODEL_ID = "us.amazon.nova-lite-v1:0";
    private static final HttpClient HTTP = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(20))
            .build();

    private AgentCore() {
    }

    public record Persona(String id, String name, String profile) {
    }

    public static Persona personaById(String personaId) {
        for (Persona p : PERSONAS) {
            if (p.id().equals(personaId)) {
                return p;
            }
        }
        return null;
    }

    private static volatile String modeOverride = null;

    /** Runtime mode override for console (m)ode cycling; null clears. */
    public static void setModeOverride(String mode) {
        if (mode == null || mode.isBlank()) {
            modeOverride = null;
            return;
        }
        String cleaned = mode.trim().toLowerCase();
        modeOverride = List.of("stub", "ollama", "bedrock", "anthropic").contains(cleaned)
                ? cleaned
                : "stub";
    }

    public static String resolveMode() {
        String mode = (modeOverride != null ? modeOverride : env("AGENT_LLM_MODE", "stub"))
                .trim()
                .toLowerCase();
        if (List.of("stub", "ollama", "bedrock", "anthropic").contains(mode)) {
            return mode;
        }
        return "stub";
    }

    public static String providerLabel(String mode) {
        return switch (mode) {
            case "stub" -> "stub";
            case "ollama" -> "ollama";
            case "bedrock" -> "bedrock";
            case "anthropic" -> "anthropic";
            default -> mode;
        };
    }

    public static String modelLabel(String mode) {
        String override = env("AGENT_LLM_MODEL", "").trim();
        if (!override.isEmpty()) {
            return override;
        }
        return switch (mode) {
            case "stub" -> "default-no-llm";
            case "ollama" -> {
                String m = env("OLLAMA_MODEL", "llama3.2:3b").trim();
                yield m.isEmpty() ? "llama3.2:3b" : m;
            }
            case "bedrock" -> {
                String m = env("AGENT_BEDROCK_MODEL_ID", "").trim();
                yield m.isEmpty() ? DEFAULT_BEDROCK_MODEL_ID : m;
            }
            case "anthropic" -> env("ANTHROPIC_MODEL", "claude-3-haiku-20240307").trim();
            default -> "(unknown)";
        };
    }

    public static void generateStream(
            Persona persona,
            List<Map<String, Object>> tickerResults,
            Consumer<Map<String, Object>> emit
    ) {
        String mode = resolveMode();
        String provider = providerLabel(mode);
        String model = modelLabel(mode);
        String userInput = buildUserInput(tickerResults);

        Map<String, Object> meta = new LinkedHashMap<>();
        meta.put("type", "meta");
        meta.put("persona", Map.of(
                "id", persona.id(),
                "name", persona.name(),
                "profile", persona.profile()
        ));
        meta.put("input", userInput);
        meta.put("provider", provider);
        meta.put("model", model);
        meta.put("mode", mode);
        meta.put("stories", tickerResults == null ? List.of() : tickerResults);
        emit.accept(meta);

        long started = System.nanoTime();
        Map<String, Object> metrics = emptyMetrics();

        try {
            switch (mode) {
                case "stub" -> generateStub(persona, started, metrics, userInput, tickerResults, emit);
                case "ollama" -> generateOllama(
                        persona, model, started, metrics, tickerResults, userInput, emit);
                case "bedrock" -> {
                    emit.accept(Map.of(
                            "type", "error",
                            "message",
                            "Mode 'bedrock' is not wired in the Java example yet. "
                                    + "Use AGENT_LLM_MODE=stub or ollama here, "
                                    + "or run the Python web app for Bedrock."
                    ));
                    metrics.put("finish_reason", "error");
                }
                default -> {
                    emit.accept(Map.of(
                            "type", "error",
                            "message",
                            "Mode '" + mode + "' is configured but not implemented yet. "
                                    + "Use AGENT_LLM_MODE=stub or ollama."
                    ));
                    metrics.put("finish_reason", "error");
                }
            }
        } catch (Exception exc) {
            emit.accept(Map.of("type", "error", "message", String.valueOf(exc.getMessage())));
            metrics.put("finish_reason", "error");
        }

        metrics.put("latency_ms", (System.nanoTime() - started) / 1_000_000L);
        emit.accept(Map.of("type", "metrics", "metrics", metrics));
        emit.accept(Map.of("type", "done"));
    }

    private static void generateStub(
            Persona persona,
            long started,
            Map<String, Object> metrics,
            String userInput,
            List<Map<String, Object>> tickerResults,
            Consumer<Map<String, Object>> emit
    ) throws InterruptedException {
        String text = stubResponse(persona, tickerResults);
        boolean first = true;
        for (String chunk : chunkText(text, 12)) {
            if (first) {
                metrics.put("ttft_ms", (System.nanoTime() - started) / 1_000_000L);
                first = false;
            }
            emit.accept(Map.of("type", "token", "text", chunk));
            Thread.sleep(20);
        }
        metrics.put("finish_reason", "stop");
        fillTokenEstimates(text, metrics, userInput);
    }

    private static void generateOllama(
            Persona persona,
            String model,
            long started,
            Map<String, Object> metrics,
            List<Map<String, Object>> tickerResults,
            String userInput,
            Consumer<Map<String, Object>> emit
    ) throws IOException, InterruptedException {
        String host = env("OLLAMA_HOST", "http://127.0.0.1:11434").replaceAll("/$", "");
        String url = host + "/api/chat";

        List<Map<String, String>> messages = buildMessages(persona, tickerResults);
        Map<String, Object> payload = Map.of(
                "model", model,
                "stream", true,
                "messages", messages
        );
        String body = new com.google.gson.Gson().toJson(payload);

        HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                .timeout(Duration.ofSeconds(120))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();

        HttpResponse<java.io.InputStream> response;
        try {
            response = HTTP.send(request, HttpResponse.BodyHandlers.ofInputStream());
        } catch (IOException | InterruptedException exc) {
            throw new IOException(
                    "Ollama request failed (" + host + "): " + exc.getMessage()
                            + ". Is Ollama running, and is OLLAMA_MODEL pulled?",
                    exc
            );
        }
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException(
                    "Ollama request failed (" + host + "): HTTP " + response.statusCode()
                            + ". Is Ollama running, and is OLLAMA_MODEL pulled?"
            );
        }

        StringBuilder textParts = new StringBuilder();
        boolean first = true;
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(response.body(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty()) {
                    continue;
                }
                JsonObject data = JsonParser.parseString(line).getAsJsonObject();
                if (data.has("error")) {
                    throw new IOException(String.valueOf(data.get("error")));
                }
                String content = "";
                if (data.has("message") && data.get("message").isJsonObject()) {
                    JsonObject message = data.getAsJsonObject("message");
                    if (message.has("content") && !message.get("content").isJsonNull()) {
                        content = message.get("content").getAsString();
                    }
                }
                if (!content.isEmpty()) {
                    if (first) {
                        metrics.put("ttft_ms", (System.nanoTime() - started) / 1_000_000L);
                        first = false;
                    }
                    textParts.append(content);
                    emit.accept(Map.of("type", "token", "text", content));
                }
                if (data.has("done") && data.get("done").getAsBoolean()) {
                    break;
                }
            }
        }

        metrics.put("finish_reason", "stop");
        fillTokenEstimates(textParts.toString(), metrics, userInput);
    }

    private static List<Map<String, String>> buildMessages(
            Persona persona,
            List<Map<String, Object>> tickerResults
    ) {
        // persona reserved for later LaunchDarkly prompt variation
        voidUnused(persona);
        return List.of(
                Map.of("role", "system", "content", loadSystemPrompt()),
                Map.of("role", "user", "content", buildUserInput(tickerResults))
        );
    }

    private static String buildUserInput(List<Map<String, Object>> tickerResults) {
        if (tickerResults == null || tickerResults.isEmpty()) {
            return CANNED_INPUT;
        }
        return YahooNews.formatStoriesForPrompt(tickerResults);
    }

    private static String loadSystemPrompt() {
        Path path = resolveSystemPromptPath();
        try {
            String text = Files.readString(path, StandardCharsets.UTF_8).trim();
            if (text.isEmpty()) {
                throw new IllegalStateException("System prompt file is empty: " + path);
            }
            return text;
        } catch (IOException exc) {
            throw new IllegalStateException(
                    "Could not read system prompt at " + path + ": " + exc.getMessage(), exc);
        }
    }

    private static Path resolveSystemPromptPath() {
        Path cwd = Path.of("").toAbsolutePath().normalize();
        Path candidate = cwd.resolve("../prompts/system_prompt.txt").normalize();
        if (Files.isRegularFile(candidate)) {
            return candidate;
        }
        candidate = cwd.resolve("prompts/system_prompt.txt").normalize();
        if (Files.isRegularFile(candidate)) {
            return candidate;
        }
        // Running from repo root: 01-reference-agent/prompts/...
        candidate = cwd.resolve("01-reference-agent/prompts/system_prompt.txt").normalize();
        if (Files.isRegularFile(candidate)) {
            return candidate;
        }
        return cwd.resolve("../prompts/system_prompt.txt").normalize();
    }

    private static String stubResponse(Persona persona, List<Map<String, Object>> tickerResults) {
        StringBuilder lines = new StringBuilder();
        lines.append("[stub / default-no-llm]\n");
        lines.append("Persona: ").append(persona.name()).append(" (").append(persona.profile()).append(")\n\n");
        lines.append("Headline briefing (stub):\n");
        if (tickerResults == null || tickerResults.isEmpty()) {
            lines.append("- (no stories loaded — click Get Stories)\n");
        } else {
            for (Map<String, Object> block : tickerResults) {
                String ticker = String.valueOf(block.getOrDefault("ticker", "?"));
                lines.append("- ").append(ticker).append(":\n");
                @SuppressWarnings("unchecked")
                List<Map<String, Object>> stories = (List<Map<String, Object>>) block.get("stories");
                if (stories == null || stories.isEmpty()) {
                    lines.append("  (no stories)\n");
                } else {
                    for (Map<String, Object> story : stories) {
                        lines.append("  • ")
                                .append(String.valueOf(story.getOrDefault("title", "(untitled)")))
                                .append("\n");
                    }
                }
            }
        }
        lines.append("\nAs a ").append(persona.profile())
                .append(" analyst, this is boilerplate report text for UI testing. ")
                .append("Switch AGENT_LLM_MODE to ollama or bedrock for a real model response.");
        return lines.toString();
    }

    private static List<String> chunkText(String text, int size) {
        List<String> chunks = new ArrayList<>();
        for (int i = 0; i < text.length(); i += size) {
            chunks.add(text.substring(i, Math.min(text.length(), i + size)));
        }
        return chunks;
    }

    private static Map<String, Object> emptyMetrics() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("latency_ms", null);
        m.put("ttft_ms", null);
        m.put("prompt_tokens", null);
        m.put("completion_tokens", null);
        m.put("total_tokens", null);
        m.put("finish_reason", null);
        return m;
    }

    private static void fillTokenEstimates(String completionText, Map<String, Object> metrics, String userInput) {
        int prompt = estimateTokens(loadSystemPrompt() + userInput);
        int completion = estimateTokens(completionText);
        metrics.put("prompt_tokens", prompt);
        metrics.put("completion_tokens", completion);
        metrics.put("total_tokens", prompt + completion);
    }

    private static int estimateTokens(String text) {
        if (text == null || text.isEmpty()) {
            return 1;
        }
        return Math.max(1, text.length() / 4);
    }

    private static String env(String key, String fallback) {
        String value = System.getenv(key);
        return value == null ? fallback : value;
    }

    private static void voidUnused(Object ignored) {
        // reserved for future per-persona / LD prompt selection
    }
}

File: 01-reference-agent/dotnet/AgentCore.cs
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace ReferenceAgent;

/// <summary>
/// Domain logic for 01-reference-agent (no HTTP here, no LaunchDarkly).
///
/// =============================================================================
/// HOW TO READ THIS FILE
/// =============================================================================
///
/// This is the **baseline** the 20-agent-config/2x .NET ports build on. Everything
/// that later becomes a LaunchDarkly AgentControl evaluation is, here, just a file
/// read and an environment variable:
///
///   1. Data          Personas (UI labels only — no LD context needed yet)
///   2. Config        Resolve provider mode + model labels from env vars
///   3. Prompting     Build chat messages (system = ../prompts/system_prompt.txt,
///                    user = stories)
///   4. Generation    GenerateStreamAsync() — the main orchestration loop
///   5. Providers     Stub (default), Ollama (local) — Bedrock is a documented
///                    "not wired" error here, matching the Node.js port
///
/// Where LaunchDarkly goes next: compare this file to
/// 20-agent-config/21-agent-completion-config/dotnet/AgentCore.cs — that port
/// replaces steps 2/3 with an `LdAiClient.CompletionConfig(...)` evaluation
/// (AgentControl). See https://launchdarkly.com/docs/home/agentcontrol for the
/// capability this baseline is standing in for.
///
/// Event contract (mirrors agentCore.js): GenerateStreamAsync yields, in order:
///   meta     — once at start: persona, input, provider, model, mode, stories
///   token    — zero or more times: streamed text fragments
///   error    — optional: human-readable failure for the Status panel
///   metrics  — once near the end: latency, tokens, finish_reason, … (snake_case)
///   done     — once at the very end: stream complete
/// </summary>
public static class AgentCore
{
    /// <summary>Selectable demo identity — a UI label only in this baseline.</summary>
    public sealed record Persona(string Id, string Name, string Profile);

    public static readonly IReadOnlyList<Persona> Personas = new List<Persona>
    {
        new("conservative-charlie", "Conservative Charlie", "conservative"),
        new("neutral-nancy", "Neutral Nancy", "neutral"),
        new("thoughtless-toby", "Thoughtless Toby", "risk-taker"),
    };

    private const string CannedInput =
        "No ticker stories loaded yet. Ask the user to click Get Stories, " +
        "then produce a brief placeholder note that you are waiting for headlines.";

    // Sensible Bedrock defaults when env vars are incomplete (kept for label parity
    // with the Node/Python ports even though Bedrock generation is not wired here).
    private const string DefaultBedrockModelId = "us.amazon.nova-lite-v1:0";
    private const string DefaultAwsRegion = "us-east-1";
    private const string DefaultAwsProfile = "Administrator";

    private static readonly string[] KnownModes = { "stub", "ollama", "bedrock", "anthropic" };

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(120) };

    public static Persona? PersonaById(string id) => Personas.FirstOrDefault(p => p.Id == id);

    private static string SystemPromptPath() => Path.Combine(YahooNews.ExampleRoot(), "prompts", "system_prompt.txt");

    /// <summary>
    /// Read the shared system prompt from the repository prompt file.
    /// Re-reads on each call so prompt edits apply without restarting the server.
    /// </summary>
    public static string LoadSystemPrompt()
    {
        var path = SystemPromptPath();
        string text;
        try
        {
            text = File.ReadAllText(path).Trim();
        }
        catch (Exception exc)
        {
            throw new InvalidOperationException($"Could not read system prompt at {path}: {exc.Message}", exc);
        }
        if (text.Length == 0)
        {
            throw new InvalidOperationException($"System prompt file is empty: {path}");
        }
        return text;
    }

    /// <summary>
    /// AGENT_LLM_MODE, defaulting to stub for zero-credential demos.
    /// Supported today: stub, ollama. bedrock is a documented "not wired" error;
    /// anthropic is reserved for later wiring (matches the Node.js port).
    /// </summary>
    public static string ResolveMode()
    {
        var mode = (Environment.GetEnvironmentVariable("AGENT_LLM_MODE") ?? "stub").Trim().ToLowerInvariant();
        return KnownModes.Contains(mode) ? mode : "stub";
    }

    public static string ResolveAwsRegion()
        => (Environment.GetEnvironmentVariable("AWS_REGION") ?? "").Trim() is { Length: > 0 } r ? r
            : (Environment.GetEnvironmentVariable("AWS_DEFAULT_REGION") ?? "").Trim() is { Length: > 0 } d ? d
            : DefaultAwsRegion;

    public static string ResolveAwsProfile()
        => (Environment.GetEnvironmentVariable("AWS_PROFILE") ?? "").Trim() is { Length: > 0 } p ? p : DefaultAwsProfile;

    /// <summary>Short provider name shown next to the model in the UI.</summary>
    public static string ProviderLabel(string mode) => mode switch
    {
        "stub" => "stub",
        "ollama" => "ollama",
        "bedrock" => "bedrock",
        "anthropic" => "anthropic",
        _ => mode,
    };

    public static string DefaultOllamaModel()
        => (Environment.GetEnvironmentVariable("OLLAMA_MODEL") ?? "").Trim() is { Length: > 0 } m ? m : "llama3.2:3b";

    /// <summary>
    /// Model id / display name for the Provider / model panel.
    /// Precedence: AGENT_LLM_MODEL override (any mode) → mode-specific defaults / env vars.
    /// </summary>
    public static string ModelLabel(string mode)
    {
        var override_ = (Environment.GetEnvironmentVariable("AGENT_LLM_MODEL") ?? "").Trim();
        if (override_.Length > 0) return override_;
        return mode switch
        {
            "stub" => "default-no-llm",
            "ollama" => DefaultOllamaModel(),
            "bedrock" => (Environment.GetEnvironmentVariable("AGENT_BEDROCK_MODEL_ID") ?? "").Trim() is { Length: > 0 } b
                ? b
                : DefaultBedrockModelId,
            "anthropic" => (Environment.GetEnvironmentVariable("ANTHROPIC_MODEL") ?? "").Trim() is { Length: > 0 } a
                ? a
                : "claude-3-haiku-20240307",
            _ => "(unknown)",
        };
    }

    private static string BuildUserInput(List<Dictionary<string, object?>>? tickerResults)
        => tickerResults is { Count: > 0 } ? YahooNews.FormatStoriesForPrompt(tickerResults) : CannedInput;

    /// <summary>
    /// Chat messages sent to real LLM providers.
    /// system ← ../prompts/system_prompt.txt (shared; LaunchDarkly later)
    /// user   ← story-based briefing prompt (same story set for every persona)
    /// </summary>
    private static List<Dictionary<string, string>> BuildMessages(List<Dictionary<string, object?>>? tickerResults) => new()
    {
        new() { ["role"] = "system", ["content"] = LoadSystemPrompt() },
        new() { ["role"] = "user", ["content"] = BuildUserInput(tickerResults) },
    };

    private static int EstimateTokens(string? text) => Math.Max(1, (text ?? "").Length / 4);

    private static Dictionary<string, object?> PersonaMap(Persona persona) => new()
    {
        ["id"] = persona.Id,
        ["name"] = persona.Name,
        ["profile"] = persona.Profile,
    };

    private sealed class Metrics
    {
        public long? LatencyMs;
        public long? TtftMs;
        public int? PromptTokens;
        public int? CompletionTokens;
        public int? TotalTokens;
        public string? FinishReason;

        public Dictionary<string, object?> ToMap() => new()
        {
            ["latency_ms"] = LatencyMs,
            ["ttft_ms"] = TtftMs,
            ["prompt_tokens"] = PromptTokens,
            ["completion_tokens"] = CompletionTokens,
            ["total_tokens"] = TotalTokens,
            ["finish_reason"] = FinishReason,
        };
    }

    private static void FillTokenEstimates(string completionText, Metrics metrics, string userInput)
    {
        metrics.PromptTokens = EstimateTokens(LoadSystemPrompt() + userInput);
        metrics.CompletionTokens = EstimateTokens(completionText);
        metrics.TotalTokens = (metrics.PromptTokens ?? 0) + (metrics.CompletionTokens ?? 0);
    }

    private static IEnumerable<string> ChunkText(string text, int size = 12)
    {
        for (var i = 0; i < text.Length; i += size)
        {
            yield return text.Substring(i, Math.Min(size, text.Length - i));
        }
    }

    /// <summary>Boilerplate text for AGENT_LLM_MODE=stub (default-no-llm).</summary>
    private static string StubResponse(Persona persona, List<Dictionary<string, object?>>? tickerResults)
    {
        var lines = new List<string>
        {
            "[stub / default-no-llm]",
            $"Persona: {persona.Name} ({persona.Profile})",
            "",
            "Headline briefing (stub):",
        };
        if (tickerResults is not { Count: > 0 })
        {
            lines.Add("- (no stories loaded — click Get Stories)");
        }
        else
        {
            foreach (var block in tickerResults)
            {
                var ticker = block.GetValueOrDefault("ticker") as string ?? "?";
                lines.Add($"- {ticker}:");
                var stories = block.GetValueOrDefault("stories") as List<Dictionary<string, object?>> ?? new();
                if (stories.Count == 0) lines.Add("  (no stories)");
                foreach (var story in stories)
                {
                    lines.Add($"  \u2022 {story.GetValueOrDefault("title") as string ?? "(untitled)"}");
                }
            }
        }
        lines.Add("");
        lines.Add(
            $"As a {persona.Profile} analyst, this is boilerplate report text for UI testing. " +
            "Switch AGENT_LLM_MODE to ollama or bedrock for a real model response.");
        return string.Join("\n", lines);
    }

    private static async IAsyncEnumerable<Dictionary<string, object?>> GenerateStubAsync(
        Persona persona, Stopwatch sw, Metrics metrics, string userInput,
        List<Dictionary<string, object?>>? tickerResults, [EnumeratorCancellation] CancellationToken ct)
    {
        var text = StubResponse(persona, tickerResults);
        var first = true;
        foreach (var chunk in ChunkText(text, 12))
        {
            if (first)
            {
                metrics.TtftMs = (long)sw.Elapsed.TotalMilliseconds;
                first = false;
            }
            yield return new Dictionary<string, object?> { ["type"] = "token", ["text"] = chunk };
            await Task.Delay(20, ct);
        }
        metrics.FinishReason = "stop";
        FillTokenEstimates(text, metrics, userInput);
    }

    private static async IAsyncEnumerable<string> OllamaStreamAsync(
        string model, List<Dictionary<string, string>> messages, [EnumeratorCancellation] CancellationToken ct)
    {
        var host = (Environment.GetEnvironmentVariable("OLLAMA_HOST") ?? "http://127.0.0.1:11434").TrimEnd('/');
        var url = $"{host}/api/chat";

        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new StringContent(
                JsonSerializer.Serialize(new { model, stream = true, messages }),
                Encoding.UTF8, "application/json"),
        };

        HttpResponseMessage response;
        try
        {
            response = await Http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);
        }
        catch (Exception exc)
        {
            throw new InvalidOperationException(
                $"Ollama request failed ({host}): {exc.Message}. " +
                "Is Ollama running, and is OLLAMA_MODEL pulled?", exc);
        }

        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException(
                $"Ollama request failed ({host}): HTTP {(int)response.StatusCode}. " +
                "Is Ollama running, and is OLLAMA_MODEL pulled?");
        }

        await using var stream = await response.Content.ReadAsStreamAsync(ct);
        using var reader = new StreamReader(stream, Encoding.UTF8);
        while (await reader.ReadLineAsync(ct) is { } line)
        {
            if (string.IsNullOrWhiteSpace(line)) continue;

            using var doc = JsonDocument.Parse(line);
            var root = doc.RootElement;
            if (root.TryGetProperty("error", out var errEl))
            {
                throw new InvalidOperationException(errEl.ToString());
            }

            var content = "";
            if (root.TryGetProperty("message", out var msgEl) &&
                msgEl.TryGetProperty("content", out var contentEl) &&
                contentEl.ValueKind == JsonValueKind.String)
            {
                content = contentEl.GetString() ?? "";
            }
            if (content.Length > 0) yield return content;

            if (root.TryGetProperty("done", out var doneEl) && doneEl.ValueKind == JsonValueKind.True) yield break;
        }
    }

    /// <summary>
    /// Stream tokens from Ollama, updating <paramref name="metrics"/> in place and yielding
    /// "token"/"error" events. Kept as its own async iterator (rather than inlined in
    /// GenerateStreamAsync) because C# disallows `yield return` inside a try block that has a
    /// catch clause — the inner try/catch below only ever *sets* a local, it never yields.
    /// </summary>
    private static async IAsyncEnumerable<Dictionary<string, object?>> GenerateOllamaAsync(
        string model, List<Dictionary<string, string>> messages, Stopwatch sw, Metrics metrics,
        string userInput, [EnumeratorCancellation] CancellationToken ct)
    {
        var textParts = new StringBuilder();
        var first = true;
        var enumerator = OllamaStreamAsync(model, messages, ct).GetAsyncEnumerator(ct);
        try
        {
            while (true)
            {
                var hasNext = false;
                string? chunk = null;
                Exception? error = null;
                try
                {
                    hasNext = await enumerator.MoveNextAsync();
                    if (hasNext) chunk = enumerator.Current;
                }
                catch (Exception exc)
                {
                    error = exc;
                }

                if (error != null)
                {
                    yield return new Dictionary<string, object?> { ["type"] = "error", ["message"] = error.Message };
                    metrics.FinishReason = "error";
                    yield break;
                }
                if (!hasNext) break;

                if (first)
                {
                    metrics.TtftMs = (long)sw.Elapsed.TotalMilliseconds;
                    first = false;
                }
                textParts.Append(chunk);
                yield return new Dictionary<string, object?> { ["type"] = "token", ["text"] = chunk };
            }
        }
        finally
        {
            await enumerator.DisposeAsync();
        }

        metrics.FinishReason = "stop";
        FillTokenEstimates(textParts.ToString(), metrics, userInput);
    }

    /// <summary>
    /// Run one generation and yield UI events in order (meta → token* → error? → metrics → done).
    ///
    /// This is the seam later .NET ports (20-agent-config/2x) replace with a LaunchDarkly
    /// AgentControl evaluation — everything above "Step 2" here becomes a config lookup there.
    /// </summary>
    public static async IAsyncEnumerable<Dictionary<string, object?>> GenerateStreamAsync(
        Persona persona,
        List<Dictionary<string, object?>>? tickerResults,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var mode = ResolveMode();
        var provider = ProviderLabel(mode);
        var model = ModelLabel(mode);
        var userInput = BuildUserInput(tickerResults);
        var stories = tickerResults ?? new List<Dictionary<string, object?>>();

        // Step 1 — describe the run up front.
        yield return new Dictionary<string, object?>
        {
            ["type"] = "meta",
            ["persona"] = PersonaMap(persona),
            ["input"] = userInput,
            ["provider"] = provider,
            ["model"] = model,
            ["mode"] = mode,
            ["stories"] = stories,
        };

        var sw = Stopwatch.StartNew();
        var metrics = new Metrics();

        // Step 2 — call the provider and stream tokens (or surface an error).
        if (mode == "stub")
        {
            await foreach (var evt in GenerateStubAsync(persona, sw, metrics, userInput, tickerResults, ct))
            {
                yield return evt;
            }
        }
        else if (mode == "ollama")
        {
            var messages = BuildMessages(tickerResults);
            await foreach (var evt in GenerateOllamaAsync(model, messages, sw, metrics, userInput, ct))
            {
                yield return evt;
            }
        }
        else if (mode == "bedrock")
        {
            yield return new Dictionary<string, object?>
            {
                ["type"] = "error",
                ["message"] =
                    "Mode 'bedrock' is not wired in the .NET example yet. " +
                    "Use AGENT_LLM_MODE=stub or ollama here, or run the Python web app for Bedrock.",
            };
            metrics.FinishReason = "error";
        }
        else
        {
            yield return new Dictionary<string, object?>
            {
                ["type"] = "error",
                ["message"] =
                    $"Mode '{mode}' is configured but not implemented in this reference yet. " +
                    "Use AGENT_LLM_MODE=stub or ollama.",
            };
            metrics.FinishReason = "error";
        }

        // Step 3 — always close with metrics + done so the UI can re-enable buttons.
        metrics.LatencyMs = (long)sw.Elapsed.TotalMilliseconds;
        yield return new Dictionary<string, object?> { ["type"] = "metrics", ["metrics"] = metrics.ToMap() };
        yield return new Dictionary<string, object?> { ["type"] = "done" };
    }
}
